BO notebook (issue #26)

NoemieJaquier opened the pull request:
This PR adds an example of BO on the sphere manifold. It also fixes batch issues for the sphere kernel (the batch computation is needed for BOtorch).
Fixes issue #26.

vabor112 reviewed (approved):
I pushed some minor changes to the text parts of the notebook. Also, I added the cell outputs to the notebook, which were missing.

A review from @stoprightthere would also be very much appreciated. Notice the warnings in the cell outputs, should we do something about them?

stoprightthere commented:
I'm on it

stoprightthere commented:
Overall LGTM.

I'd want to see outputs (plots) in cells 12 to the last cell.

Another (minor) thing is that `ackley_function_sphere` is written in numpy, essentially. Is it too painful to rewrite it in torch? (probably it will need setting environment vars like `GEOMSTATS_BACKEND=<backend_name>`) 
It is not an issue _per se_ but if there are some issues with this, probably we best learn about them sooner than later.

As for the warnings: they look like botorch/gpytroch internal warnings so we can scarcely do anything about it. I don't see a big deal if we just suppress them.

vabor112 commented:
Added the missing cell outputs. Also suppressed the botorch/gpytroch internal warnings and fixed other warnings.

vabor112 commented:
@NoemieJaquier @stoprightthere should I merge this one or we want to do something about

> Another (minor) thing is that ackley_function_sphere is written in numpy, essentially. Is it too painful to rewrite it in torch? (probably it will need setting environment vars like GEOMSTATS_BACKEND=<backend_name>)
It is not an issue per se but if there are some issues with this, probably we best learn about them sooner than later.

stoprightthere commented:
Interesting to hear @NoemieJaquier 's comments on this

NoemieJaquier commented:
I agree, it is better if the Ackley function is implemented in torch. I updated this. I also added our own implementation of the logarithmic map for the sphere -- turns out that the geomstats "to_tangent" is not the logmap but the projection on the tangent space, which was not adapted for what we need here. 
I also fixed couple of typos.
@stoprightthere @vabor112 May you double check that all is fine for you? If yes, I'll merge.

vabor112 commented on notebooks/bo_sphere.ipynb:
```diff
@@ -108,39 +109,63 @@
    "source": [
+    "def logmap(x, x0):\n",
+    "    \"\"\"\n",
+    "    This functions maps a point lying on the sphere manifold into the tangent space of a second point of the manifold.\n",
```
Probably
- "this function" instead of "this functions"
- "on the manifold" instead of "of the manifold"

NoemieJaquier replied:
Done, thanks!

vabor112 commented on notebooks/bo_sphere.ipynb:
```diff
@@ -108,39 +109,63 @@
-    "    # The first coordinate of x_proj is always 0, so that vectors in the tangent space can be expressed in a\n",
-    "    # dimension-1 dimensional space by simply ignoring the first coordinate.\n",
-    "    base = np.zeros((1, dimension))\n",
+    "    # The first coordinate of x_proj is always 0, so that vectors in the tangent space can be expressed in a dim-1\n",
```
Maybe write "\`dimension-1\`" instead of "dim-1"? The variable we are talking about is called `dimension`, not `dim`.

vabor112 commented on notebooks/bo_sphere.ipynb:
```diff
@@ -194,23 +219,31 @@
+    "Defining a Gaussian process model also requires a likelihood noise hyperparameter $\\sigma_n^2$.\n",
     "\n",
-    "We use the $\\operatorname{Gamma}(2.0, 0.15)$ prior for $\\sigma$, the squre root of $\\sigma^2$, and $\\operatorname{Gamma}(1.1, 0.05)$ prior for $\\sigma_n^2$."
+    "We use the $\\operatorname{Gamma}(2.0, 0.15)$ prior for $\\sigma$, the square root of $\\sigma^2$, and $\\operatorname{Gamma}(1.1, 0.05)$ prior for $\\sigma_n^2$."
```
Lol, the density of typos here is impressive. My mind must have been partly elsewhere when I wrote this :)

stoprightthere commented:
LGTM
